### PR TITLE
Use extension for positional record checks

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Rules/GenericInheritanceShouldNotBeRecursiveBase.cs [hunk 1]
```diff
@@ -21,6 +21,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Extensions;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/GenericInheritanceShouldNotBeRecursiveBase.cs:24]: `MethodsShouldNotHaveIdenticalImplementationsBase` should be also updated to remove `IsExcludedFromBeingExamined`

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposableTypesNeedFinalizersTest.cs [hunk 2]
```diff
@@ -25,22 +25,31 @@ namespace SonarAnalyzer.UnitTest.Rules
     [TestClass]
     public class DisposableTypesNeedFinalizersTest
     {
+        private readonly VerifierBuilder builder = new VerifierBuilder<DisposableTypesNeedFinalizers>();
+
         [TestMethod]
         public void DisposableTypesNeedFinalizers() =>
-            OldVerifier.VerifyAnalyzer(@"TestCases\DisposableTypesNeedFinalizers.cs", new DisposableTypesNeedFinalizers());
+            builder.AddPaths("DisposableTypesNeedFinalizers.cs")
+                .Verify();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposableTypesNeedFinalizersTest.cs:33]: Fits on single line

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposableTypesNeedFinalizersTest.cs [hunk 3]
```diff
@@ -25,22 +25,31 @@ namespace SonarAnalyzer.UnitTest.Rules
...
 
 #if NET
+
         [TestMethod]
         public void DisposableTypesNeedFinalizers_CSharp9() =>
-            OldVerifier.VerifyAnalyzerFromCSharp9Library(@"TestCases\DisposableTypesNeedFinalizers.CSharp9.cs", new DisposableTypesNeedFinalizers());
+            builder.AddPaths("DisposableTypesNeedFinalizers.CSharp9.cs")
+                .WithOptions(ParseOptionsHelper.FromCSharp9)
+                .Verify();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposableTypesNeedFinalizersTest.cs:41]: Also can fit on a single line

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotDecreaseMemberVisibilityTest.cs [hunk 4]
```diff
@@ -38,7 +38,9 @@ public class DoNotDecreaseMemberVisibilityTest
 
         [TestMethod]
         public void DoNotDecreaseMemberVisibility_CSharp9() =>
-            builder.AddPaths("DoNotDecreaseMemberVisibility.CSharp9.cs").WithOptions(ParseOptionsHelper.FromCSharp9).Verify();
+            builder.AddPaths("DoNotDecreaseMemberVisibility.CSharp9.cs")
+                .WithOptions(ParseOptionsHelper.FromCSharp9)
+                .Verify();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotDecreaseMemberVisibilityTest.cs:43]: This fitted nicely on a single line before.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotHideBaseClassMethodsTest.cs [hunk 5]
```diff
@@ -25,14 +25,25 @@ namespace SonarAnalyzer.UnitTest.Rules
...
     public class DoNotHideBaseClassMethodsTest
     {
+        private readonly VerifierBuilder builder = new VerifierBuilder<DoNotHideBaseClassMethods>();
+
         [TestMethod]
         public void DoNotHideBaseClassMethods() =>
-            new VerifierBuilder<DoNotHideBaseClassMethods>().AddPaths("DoNotHideBaseClassMethods.cs", "DoNotHideBaseClassMethods2.cs").WithAutogenerateConcurrentFiles(false).Verify();
+            builder.AddPaths(
+                "DoNotHideBaseClassMethods.cs",
+                "DoNotHideBaseClassMethods2.cs")
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotHideBaseClassMethodsTest.cs:34]: I generally disagree with adding useless EOLS in all UTs.
Even this formatting wastes +2 for arguments that are not long.
```suggestion
            builder.AddPaths("DoNotHideBaseClassMethods.cs", "DoNotHideBaseClassMethods2.cs")
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DisposableTypesNeedFinalizers.cs [hunk 6]
```diff
@@ -35,28 +36,22 @@ public sealed class DisposableTypesNeedFinalizers : SonarDiagnosticAnalyzer
         private const string DiagnosticId = "S4002";
         private const string MessageFormat = "Implement a finalizer that calls your 'Dispose' method.";
 
-        private static readonly DiagnosticDescriptor Rule
-            = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DisposableTypesNeedFinalizers.cs:39]: The empty line is not needed after this - it's a block of static fields.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposableTypesNeedFinalizersTest.cs [hunk 7]
```diff
@@ -25,22 +25,31 @@ namespace SonarAnalyzer.UnitTest.Rules
...
         [TestMethod]
         public void DisposableTypesNeedFinalizers_InvalidCode() =>
-            OldVerifier.VerifyCSharpAnalyzer(@"
+            builder.AddSnippet(@"
 public class Foo_05 : IDisposable
 {
     private HandleRef;
-}", new DisposableTypesNeedFinalizers(), CompilationErrorBehavior.Ignore);
+}")
+                .WithErrorBehavior(CompilationErrorBehavior.Ignore)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposableTypesNeedFinalizersTest.cs:52]: This is an exception where it makes sense to have multiple lines. And I would also not complain about
```
}").WithErrorBehavior(...).Verify();`
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DisposableTypesNeedFinalizers.cs [hunk 8]
```diff
@@ -35,28 +36,22 @@ public sealed class DisposableTypesNeedFinalizers : SonarDiagnosticAnalyzer
...
-                    return;
-                }
-
                 var declaration = (TypeDeclarationSyntax)c.Node;
-                var classSymbol = c.SemanticModel.GetDeclaredSymbol(declaration);
-
-                if (classSymbol.Implements(KnownType.System_IDisposable)
+                if (!c.IsRedundantPositionalRecordContext()
+                    && c.SemanticModel.GetDeclaredSymbol(declaration) is var classSymbol
+                    && classSymbol.Implements(KnownType.System_IDisposable)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DisposableTypesNeedFinalizers.cs:54]: We already do have that symbol as input so we don't need to re-query it
```suggestion
                    && ((ITypeSymbol)c.ContainingSymbol).Implements(KnownType.System_IDisposable)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs [hunk 9]
```diff
@@ -45,7 +46,8 @@ public sealed class DoNotDecreaseMemberVisibility : SonarDiagnosticAnalyzer
                     var classDeclaration = (TypeDeclarationSyntax)c.Node;
 
                     if (classDeclaration.Identifier.IsMissing
-                        || !(c.ContainingSymbol is ITypeSymbol {Kind: SymbolKind.NamedType, IsSealed: false} classSymbol))
+                        || c.IsRedundantPositionalRecordContext()
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs:49]: This class is ugly all the way from top to bottom (`ForEach` should be `foreach(...)`, emty lines, null checks, some single-use renamings in variables, condition positivity, control flow). Not sure if it's worth fixing in this PR.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotHideBaseClassMethods.cs [hunk 10]
```diff
@@ -25,6 +25,7 @@
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Extensions;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotHideBaseClassMethods.cs:28]: Same as `DoNotDecreateMemberVisibility`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs [hunk 11]
```diff
@@ -42,7 +43,9 @@ public sealed class NonDerivedPrivateClassesShouldBeSealed : SonarDiagnosticAnal
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
             {
                 var typeDeclarationSyntax = (TypeDeclarationSyntax)c.Node;
-                if (IsPrivateButNotSealedType(typeDeclarationSyntax) && !HasVirtualMembers(typeDeclarationSyntax))
+                if (IsPrivateButNotSealedType(typeDeclarationSyntax)
+                    && !HasVirtualMembers(typeDeclarationSyntax)
+                    && !c.IsRedundantPositionalRecordContext())
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs:48]: _Performance:_ The redundant stuff should go first. The other two conditions are expensive compared to it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantModifier.cs [hunk 12]
```diff
@@ -204,8 +205,8 @@ private static void CheckTypeDeclarationForRedundantPartial(SyntaxNodeAnalysisCo
         {
             var typeDeclaration = (TypeDeclarationSyntax)context.Node;
             if (typeDeclaration.Modifiers.Any(SyntaxKind.PartialKeyword)
-                && context.ContainingSymbol.Kind == SymbolKind.NamedType
-                && context.SemanticModel.GetDeclaredSymbol(typeDeclaration) is { DeclaringSyntaxReferences: { Length: <= 1 } })
+                && !context.IsRedundantPositionalRecordContext()
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantModifier.cs:208]: _Performance:_ This should be unmeasurably better than the `Any` collection iteration check above.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs [hunk 13]
```diff
@@ -53,7 +53,7 @@ public sealed class TestClassShouldHaveTestMethod : SonarDiagnosticAnalyzer
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
-                    if (c.ContainingSymbol.Kind != SymbolKind.NamedType)
+                    if (c.IsRedundantPositionalRecordContext())
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/TestClassShouldHaveTestMethod.cs:56]: `c =>` can be on previous line. 
With moving the cast before the `if`, we can merge all 3 conditions

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Bunch of minor improvements

File: analyzers/src/SonarAnalyzer.Common/Rules/GenericInheritanceShouldNotBeRecursiveBase.cs (hunk 1 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/GenericInheritanceShouldNotBeRecursiveBase.cs:24]: Here, the check is more than just the same is redundant positional record check, it plays a role also for checking namespaces for top level statements.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs (hunk 9 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs:49]: I did an additional minimal cleanup, the other parts I would leave for future work on the rule.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**